runtime/caml/camlatomic.h

```diff
@@ -0,0 +1,87 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*      KC Sivaramakrishnan, Indian Institute of Technology, Madras       */
+/*                Stephen Dolan, University of Cambridge                  */
+/*                                                                        */
+/*   Copyright 2018 Indian Institute of Technology, Madras                */
+/*   Copyright 2018 University of Cambridge                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+#ifndef CAML_ATOMIC_H
+#define CAML_ATOMIC_H
+
+#ifndef CAML_NAME_SPACE
+#include "compatibility.h"
+#endif
+#include "config.h"
+#include "misc.h"
+
+/* On platforms supporting C11 atomics, this file just includes <stdatomic.h>.
+
+   On other platforms, this file includes platform-specific stubs for
+   the subset of C11 atomics needed by the OCaml runtime
+ */
+
+#ifdef __cplusplus
+
+extern "C++" {
+#include <atomic>
+#define ATOMIC_UINTNAT_INIT(x) (x)
+typedef std::atomic<uintnat> atomic_uintnat;
+typedef std::atomic<intnat> atomic_intnat;
+using std::memory_order_relaxed;
+using std::memory_order_acquire;
+using std::memory_order_release;
+using std::memory_order_acq_rel;
+using std::memory_order_seq_cst;
+}
+
+#elif defined(HAS_STDATOMIC_H)
+
+#include <stdatomic.h>
+#define ATOMIC_UINTNAT_INIT(x) (x)
+typedef _Atomic uintnat atomic_uintnat;
+typedef _Atomic intnat atomic_intnat;
+
+#elif defined(__GNUC__)
+
+/* Support for versions of gcc which have built-in atomics but do not
+   expose stdatomic.h (e.g. gcc 4.8) */
```

Comment:
Do we really need  to support GCC 4.8?  (last released in 2015...) 

It would be nice to rely on `<stdatomic.h>` unless absolutely impossible (i.e. unless MSVC).

Reply:
Here is a neat table of which distro ships with which compiler version: https://repology.org/project/gcc/versions

However it is possible to obtain newer versions of GCC even on some older distros sometimes.

Just one datapoint from my usecase:
E.g. the platform I work on is based on CentOS 7 (EOL in 2024) and ships with 4.8, but it is possible to install newer versions of GCC from SCL (that do not conflict with the system compiler): https://access.redhat.com/documentation/en-us/red_hat_developer_toolset/10/html-single/10.0_release_notes/index
In fact we already had to do that for another package that required a newer version of GCC (IIRC it was grub).
As long as the minimum version of GCC is clearly documented I don't see this as a major problem. 
(upgrading the compiler version used for OCaml will likely have a ripple effect on every C stub we build, as new compilers may introduce new warnings, or new bugs, etc. but we can test the compiler upgrade independently of the multicore ocaml upgrade, and cleaning up those warnings would be beneficial anyway)

Reply:
Tracking this for post-merge fixing in https://github.com/ocaml/ocaml/issues/10861